File: Java_client/bim-portal-enhanced/src/main/java/com/pb40/bimportal/examples/TemplateExportExample.java
package com.pb40.bimportal.examples;

import com.bimportal.client.model.SimpleAiaTemplatePublicDto;
import com.pb40.bimportal.client.BimPortalClientBuilder;
import com.pb40.bimportal.client.EnhancedBimPortalClient;
import com.pb40.bimportal.config.BimPortalConfig;
import com.pb40.bimportal.util.ExportUtils;
import java.nio.file.Path;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * AIA template export examples for BIM Portal Java client.
 *
 * <p>Demonstrates AIA template export workflows in multiple formats including PDF and OpenOffice
 * with automatic file type detection.
 */
public class TemplateExportExample {

  private static final Logger logger = LoggerFactory.getLogger(TemplateExportExample.class);

  /**
   * Check if credentials are available for authentication.
   *
   * @return True if credentials are configured
   */
  private static boolean checkCredentials() {
    if (!BimPortalConfig.hasCredentials()) {
      System.out.println("============================================================");
      System.out.println("WARNING: Credentials not found in environment variables.");
      System.out.println(
          "Please set "
              + BimPortalConfig.USERNAME_ENV_VAR
              + " and "
              + BimPortalConfig.PASSWORD_ENV_VAR
              + " in .env file.");
      System.out.println("============================================================");
      return false;
    }
    return true;
  }

  /**
   * Run AIA template export examples.
   *
   * @param client Enhanced BIM Portal client
   */
  public static void runTemplateExportExamples(EnhancedBimPortalClient client) {
    System.out.println("\n============================================================");
    System.out.println("📋 AIA TEMPLATE EXPORT EXAMPLES");
    System.out.println("============================================================");

    System.out.println("\n1️⃣ Searching for available templates...");
    try {
      List<SimpleAiaTemplatePublicDto> templates = client.searchTemplates();
      if (templates.isEmpty()) {
        System.out.println("🔭 No templates found for export");
        return;
      }

      System.out.println("✅ Found " + templates.size() + " templates:");
      for (int i = 0; i < Math.min(3, templates.size()); i++) {
        SimpleAiaTemplatePublicDto template = templates.get(i);
        System.out.println(
            "   " + (i + 1) + ". " + template.getName() + " (" + template.getGuid() + ")");
      }

      SimpleAiaTemplatePublicDto selectedTemplate = templates.get(0);
      System.out.println("\n🎯 Using template: '" + selectedTemplate.getName() + "'");

      Map<String, Path> exportResults = new HashMap<>();

      System.out.println("\n2️⃣ Exporting template...");

      // PDF Export with auto-detection
      System.out.println("   📄 Exporting as PDF...");
      Optional<byte[]> pdfContent = client.exportTemplatePdf(selectedTemplate.getGuid());
      if (pdfContent.isPresent()) {
        String baseFilename = "template_pdf_" + selectedTemplate.getGuid();
        Optional<Path> pdfPath =
            ExportUtils.exportWithDetection(pdfContent.get(), baseFilename, "pdf");
        if (pdfPath.isPresent()) {
          exportResults.put("PDF", pdfPath.get());
          System.out.println("   ✅ PDF exported: " + pdfPath.get());
        } else {
          System.out.println("   ❌ PDF export failed: Could not save file");
        }
      } else {
        System.out.println("   ❌ PDF export failed: No content received");
      }

      // OpenOffice Export with auto-detection
      System.out.println("   📝 Exporting as OpenOffice...");
      Optional<byte[]> odtContent = client.exportTemplateOpenOffice(selectedTemplate.getGuid());
      if (odtContent.isPresent()) {
        String baseFilename = "template_odt_" + selectedTemplate.getGuid();
        Optional<Path> odtPath =
            ExportUtils.exportWithDetection(odtContent.get(), baseFilename, "odt");
        if (odtPath.isPresent()) {
          exportResults.put("OpenOffice", odtPath.get());
          System.out.println("   ✅ OpenOffice exported: " + odtPath.get());
        } else {
          System.out.println("   ❌ OpenOffice export failed: Could not save file");
        }
      } else {
        System.out.println("   ❌ OpenOffice export failed: No content received");
      }

      // Summary with file type information
      System.out.println("\n📈 Export Summary: " + exportResults.size() + "/2 formats successful");
      for (Map.Entry<String, Path> entry : exportResults.entrySet()) {
        String filename = entry.getValue().getFileName().toString();
        String extension = filename.substring(filename.lastIndexOf('.') + 1).toUpperCase();
        System.out.println("   ✅ " + entry.getKey() + " (" + extension + "): " + entry.getValue());
      }

    } catch (Exception e) {
      logger.error("❌ Error in template export examples", e);
      System.err.println("❌ Error in template export examples: " + e.getMessage());
    }
  }

  /** Main method to run AIA template export examples. */
  public static void main(String[] args) {
    System.out.println("======================================================================");
    System.out.println("🚀 BIM PORTAL AIA TEMPLATE EXPORT EXAMPLES");
    System.out.println("======================================================================");

    if (!checkCredentials()) {
      System.err.println("❌ Cannot run export examples without credentials");
      return;
    }

    if (!ExportUtils.isExportDirectoryWritable()) {
      System.err.println("❌ Export directory is not writable!");
      return;
    }

    System.out.println("🔧 Setting up authenticated client...");

    try {
      EnhancedBimPortalClient client = BimPortalClientBuilder.buildDefault();

      EnhancedBimPortalClient.HealthCheckResult healthCheck = client.performHealthCheck();
      System.out.println("🩺 Health check: " + healthCheck);

      if (!healthCheck.isApiAccessible()) {
        System.err.println("❌ API is not accessible. Please check configuration.");
        return;
      }

      runTemplateExportExamples(client);

      System.out.println(
          "\n======================================================================");
      System.out.println("✅ AIA TEMPLATE EXPORT EXAMPLES COMPLETE!");
      System.out.println("======================================================================");
      System.out.println(
          "📂 Check the '" + BimPortalConfig.getExportPath() + "' directory for exported files");

      client.logout();
      System.out.println("👋 Logged out and session closed.");

    } catch (Exception e) {
      logger.error("❌ Error running template export examples", e);
      System.err.println("❌ Error running examples: " + e.getMessage());
    }
  }
}

File: Python_client/examples/template_export_example.py
"""
AIA template export examples for BIM Portal Python client.

This module demonstrates AIA template export workflows in multiple formats including 
PDF and OpenOffice with automatic file type detection.
"""

import os
import logging
from pathlib import Path
from typing import Dict, Optional

from dotenv import load_dotenv

from client.auth.auth_config import BIM_PORTAL_PASSWORD_ENV_VAR, BIM_PORTAL_USERNAME_ENV_VAR
from client.auth.auth_service_impl import AuthService
from client.enhanced_bim_client import EnhancedBimPortalClient
from client.config import BIMPortalConfig
from examples.export_examples.utils.export_utils import ExportUtils

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# Load environment variables
load_dotenv()


def check_credentials() -> bool:
    """
    Check if credentials are available for authentication.
    
    Returns:
        True if credentials are configured
    """
    if not os.getenv(BIM_PORTAL_USERNAME_ENV_VAR) or not os.getenv(BIM_PORTAL_PASSWORD_ENV_VAR):
        print("=" * 60)
        print("WARNING: Credentials not found in environment variables.")
        print(f"Please set {BIM_PORTAL_USERNAME_ENV_VAR} and {BIM_PORTAL_PASSWORD_ENV_VAR} in .env file.")
        print("=" * 60)
        return False
    return True


def run_template_export_examples(client: EnhancedBimPortalClient):
    """
    Run AIA template export examples with comprehensive format support.
    
    Args:
        client: Enhanced BIM Portal client
    """
    print("\n" + "=" * 60)
    print("📋 AIA TEMPLATE EXPORT EXAMPLES")
    print("=" * 60)
    
    print("\n1️⃣ Searching for available templates...")
    try:
        templates = client.search_templates()
        if not templates:
            print("🔭 No templates found for export")
            return
        
        print(f"✅ Found {len(templates)} templates:")
        for i, template in enumerate(templates[:3], 1):
            print(f"   {i}. {template.name} ({template.guid})")
        
        selected_template = templates[0]
        print(f"\n🎯 Using template: '{selected_template.name}'")
        
        export_results: Dict[str, Optional[Path]] = {}
        
        print("\n2️⃣ Exporting template...")
        
        # PDF Export with auto-detection
        print("   📄 Exporting as PDF...")
        pdf_content = client.export_template_pdf(selected_template.guid)
        if pdf_content:
            base_filename = f"template_pdf_{selected_template.guid}"
            pdf_path = ExportUtils.export_with_detection(pdf_content, base_filename, "pdf")
            if pdf_path:
                export_results['PDF'] = pdf_path
                print(f"   ✅ PDF exported: {pdf_path}")
            else:
                print("   ❌ PDF export failed: Could not save file")
                export_results['PDF'] = None
        else:
            print("   ❌ PDF export failed: No content received")
            export_results['PDF'] = None
        
        # OpenOffice Export with auto-detection
        print("   📝 Exporting as OpenOffice...")
        odt_content = client.export_template_openoffice(selected_template.guid)
        if odt_content:
            base_filename = f"template_odt_{selected_template.guid}"
            odt_path = ExportUtils.export_with_detection(odt_content, base_filename, "odt")
            if odt_path:
                export_results['OpenOffice'] = odt_path
                print(f"   ✅ OpenOffice exported: {odt_path}")
            else:
                print("   ❌ OpenOffice export failed: Could not save file")
                export_results['OpenOffice'] = None
        else:
            print("   ❌ OpenOffice export failed: No content received")
            export_results['OpenOffice'] = None
        
        # Summary with file type information
        successful_exports = len([path for path in export_results.values() if path is not None])
        total_formats = len(export_results)
        
        print(f"\n📈 Export Summary: {successful_exports}/{total_formats} formats successful")
        
        for format_name, path in export_results.items():
            if path:
                extension = path.suffix[1:].upper() if path.suffix else "UNKNOWN"
                print(f"   ✅ {format_name} ({extension}): {path}")
            else:
                print(f"   ❌ {format_name}: Export failed")
        
        # Additional template details
        print(f"\n📋 Template Details:")
        try:
            detailed_template = client.get_template(selected_template.guid)
            if detailed_template:
                print(f"   Name: {detailed_template.name}")
                print(f"   GUID: {detailed_template.guid}")
                if hasattr(detailed_template, 'description') and detailed_template.description:
                    print(f"   Description: {detailed_template.description}")
                if hasattr(detailed_template, 'version') and detailed_template.version:
                    print(f"   Version: {detailed_template.version}")
                if hasattr(detailed_template, 'template_type') and detailed_template.template_type:
                    print(f"   Type: {detailed_template.template_type}")
            else:
                print("   Could not retrieve detailed information")
        except Exception as e:
            print(f"   Error retrieving details: {e}")
        
        if successful_exports == 0:
            print("💡 Note: Template exports may require special permissions")
        
    except Exception as e:
        logger.error("Error in template export examples", exc_info=True)
        print(f"❌ Error in template export examples: {e}")


def demonstrate_template_search_and_filtering(client: EnhancedBimPortalClient):
    """
    Demonstrate template search and filtering capabilities.
    
    Args:
        client: Enhanced BIM Portal client
    """
    print("\n3️⃣ Template search and filtering examples...")
    
    try:
        # Search all templates
        all_templates = client.search_templates()
        print(f"   📊 Total templates available: {len(all_templates)}")
        
        if all_templates:
            # Show template categories or types if available
            template_types = set()
            for template in all_templates:
                if hasattr(template, 'template_type') and template.template_type:
                    template_types.add(template.template_type)
                elif hasattr(template, 'type') and template.type:
                    template_types.add(template.type)
            
            if template_types:
                print(f"   📂 Template types found: {', '.join(template_types)}")
            
            # Show templates by name length (as an example filter)
            short_names = [t for t in all_templates if len(t.name) < 20]
            long_names = [t for t in all_templates if len(t.name) >= 20]
            
            print(f"   📝 Templates with short names (<20 chars): {len(short_names)}")
            print(f"   📜 Templates with long names (≥20 chars): {len(long_names)}")
            
            # Show newest templates (if timestamps available)
            print(f"   📅 Most recent templates:")
            for template in all_templates[:3]:
                print(f"      - {template.name}")
        
    except Exception as e:
        logger.error("Error in template search examples", exc_info=True)
        print(f"   ❌ Error in template search: {e}")


def main():
    """Main method to run AIA template export examples."""
    print("=" * 70)
    print("🚀 BIM PORTAL AIA TEMPLATE EXPORT EXAMPLES")
    print("=" * 70)
    
    if not check_credentials():
        print("❌ Cannot run export examples without credentials")
        return
    
    if not ExportUtils.is_export_directory_writable():
        print("❌ Export directory is not writable!")
        return
    
    print("🔧 Setting up authenticated client...")
    
    try:
        auth_service = AuthService()
        client = EnhancedBimPortalClient(
            auth_service=auth_service,
            base_url=BIMPortalConfig.BASE_URL
        )
        
        # Run template export examples
        run_template_export_examples(client)
        
        # Demonstrate template search and filtering
        demonstrate_template_search_and_filtering(client)
        
        print("\n" + "=" * 70)
        print("✅ AIA TEMPLATE EXPORT EXAMPLES COMPLETE!")
        print("=" * 70)
        print(f"📂 Check the '{BIMPortalConfig.EXPORT_DIRECTORY}' directory for exported files")
        
    except Exception as e:
        logger.error("Error running template export examples", exc_info=True)
        print(f"❌ Error running examples: {e}")


if __name__ == "__main__":
    main()
